security/s2a/internal/crypter/counter.go

```diff
@@ -0,0 +1,34 @@
+package crypter
+
+import (
+	"errors"
+)
+
+// counter is a 64-bit counter.
+type counter struct {
+	value         uint64
+	hasOverflowed bool
+}
+
+func newCounter(value uint64) counter {
+	return counter{value: value}
+}
+
+// getAndIncrement returns the current value of the counter and increments it.
+func (c *counter) getAndIncrement() (uint64, error) {
+	if c.hasOverflowed {
+		return 0, errors.New("counter has overflowed")
+	}
+	val := c.value
+	c.value++
+	if c.value == 0 {
```

Comment:
Yes correct. Taken from the [TLS 1.3 RFC](https://tools.ietf.org/html/rfc8446#section-5.3):
```
   A 64-bit sequence number is maintained separately for reading and
   writing records.  The appropriate sequence number is incremented by
   one after reading or writing each record.  Each sequence number is
   set to zero at the beginning of a connection and whenever the key is
   changed; the first record transmitted under a particular traffic key
   MUST use sequence number 0.

   Because the size of sequence numbers is 64-bit, they should not wrap.
   If a TLS implementation would need to wrap a sequence number, it MUST
   either rekey (Section 4.6.3) or terminate the connection.
```